### PR TITLE
Clean up tracing

### PR DESCRIPTION
**Pull request by mrBliss**

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/TxSubmission.hs [hunk 1]
```diff
@@ -32,12 +32,11 @@ localTxSubmissionServer tracer Mempool{addTxs} =
...
         traceWith tracer (condense tx)
         res <- addTxs [tx]
-        -- The 'addTxs' action returns the failing ones, whereas the protocol
-        -- returns a Maybe failure for a single tx, so we must convert here.
         case res of
-          [(_tx,rej)] -> return (Just (show rej), server)
-          _           -> return (Nothing,         server),
+          [(_tx, mbErr)] -> return (show <$> mbErr, server)
+          -- The output list of addTxs has the same length as the input list.
+          _              -> error "addTxs: unexpected result"
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/TxSubmission.hs]: This patch probably was intended to be in a separate PR. It is good but unrelated.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 2]
```diff
@@ -232,35 +214,29 @@ initInternalState NodeParams {..} = do
...
         blockFetchInterface = initBlockFetchConsensusInterface
-          (tracePrefix "ChainDB" (Nothing :: Maybe peer) tracer)
           cfg chainDB getCandidates blockFetchSize blockMatchesHeader btime
 
     return IS {..}
+  where
+    NodeParams
+      { tracers, chainDB, threadRegistry, cfg, blockFetchSize
+      , blockMatchesHeader, btime, callbacks, initState
+      } = params
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: There's nothing wrong with doing the pattern matching like this in the function arg position directly. I usually think it's clearer. And yes it's better than using `{..}`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  , blockFetchServerTracer        :: f (TraceBlockFetchServerEvent blk)
+  , txInboundTracer               :: f (TraceTxSubmissionInbound  (GenTxId blk) (GenTx blk))
+  , txOutboundTracer              :: f (TraceTxSubmissionOutbound (GenTxId blk) (GenTx blk))
+  , localTxSubmissionServerTracer :: f (TraceLocalTxSubmissionServerEvent blk)
+  , mempoolTracer                 :: f (TraceEventMempool blk)
+  , forgeTracer                   :: f (TraceForgeEvent blk)
+  }
+
+-- | A record of 'Tracer's for the node.
+type Tracers m peer blk = Tracers' peer blk (Tracer m)
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]: Why did this need to be defined in two parts? When do we need these things to be not just a `Tracer` ?

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs [hunk 4]
```diff
@@ -217,6 +216,44 @@ protocolCodecsId = ProtocolCodecs {
     , pcLocalTxSubmissionCodec = codecLocalTxSubmissionId
     }
 
+-- | A record of 'Tracer's for the different protocols.
+type ProtocolTracers m peer blk failure = ProtocolTracers' peer blk failure (Tracer m)
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: Same point/question applies here.

File: ouroboros-consensus/src/Ouroboros/Consensus/TxSubmission.hs (hunk 1 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/TxSubmission.hs]: Indeed, I'll make a separate PR for it.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs (hunk 3 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]: To avoid having to define `TraceOptions` like in https://github.com/input-output-hk/cardano-node/pull/82/files#diff-95c531b5b63fb72bec464a1f996906d6R58. Instead of having to keep these in-sync (which will annoying seeing that these repos often drift out of sync), you can just get `TraceOptions` like so:
```haskell
type TraceOptions = Tracers' () () (Const Bool)
```
and then:

```haskell
parseTraceOptions :: Parser TraceOptions
parseTraceOptions = Tracers
    <$> (Const <$> parseTraceChainSyncClient)
    <*> (Const <$> parseTraceChainSyncServer)
    <*> (Const <$> parseTraceBlockFetchDecision)
    <*> (Const <$> parseTraceBlockFetchClient)
    <*> (Const <$> parseTraceBlockFetchServer)
    <*> (Const <$> parseTraceTxInbound)
    <*> (Const <$> parseTraceTxOutbound)
    <*> (Const <$> parseTraceLocalTxSubmissionServer)
    <*> (Const <$> parseTraceMempool)
    <*> (Const <$> parseTraceForge)
```

Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]: :thinking: I was thinking to do the same in `cardano-node` but there the set of options is smaller than the set of tracers.

**Comment by mrBliss** [comment]
See https://github.com/input-output-hk/cardano-node/pull/83 for how this would be used.

**Comment by mrBliss** [comment]
@dcoutts @coot should I merge this? Is https://github.com/input-output-hk/ouroboros-network/pull/839#discussion_r308188388 clear?

File: ouroboros-consensus/src/Ouroboros/Consensus/TxSubmission.hs (hunk 1 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/TxSubmission.hs]: See #840

**Comment by coot** [comment]
Yes, thanks. I'll am able to finish now `cardano-node` side.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs (hunk 3 above)
Comment (dcoutts) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]: We should talk to @CodiePP about the config for this. I don't think we're expecting to have a whole bunch of bool options, but rather to wire this up to the switchboard, and then they have more generic config for turning on/off/details of tracers.

I think we can use types to ensure these things do not drift without needing to over-paramaterise our tracers bundle type.
